### PR TITLE
Feature/lazy resolver

### PR DESCRIPTION
**Pull request by shockey**
### Description

- Implements plan outlined in http://github.com/swagger-api/swagger-ui/issues/4248.
- Supports UI changes in editor lifecycle and validation plugins

DEPENDS on https://github.com/swagger-api/swagger-js/pull/1248.
DEPENDS on https://github.com/swagger-api/swagger-ui/pull/4249.

### How Has This Been Tested?

I've tested this locally: 
- Models and Operations update correctly when edited
- Semantic Validation resolves and caches its data outside of state as necessary

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [x] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [ ] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [x] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/plugins/validate-semantic/index.js [hunk 1]
```diff
@@ -110,3 +112,13 @@ function makeTraverseOnce(getSystem) {
     return deferred.promise.then( a => a[name] )
   }
 }
+
+function makeMemoizedResolveSubtree(system) {
+  const cacheKeymaker = (obj, path) => {
+    return `${obj.toString()} ${path.join("<>")}`
```
Comment (ponelat) [inline src/plugins/validate-semantic/index.js]: is toString done on the whole object? It might be better to use the internal hash of the object, which should be unique?

Comment (shockey) [reply src/plugins/validate-semantic/index.js]: @ponelat yeah, it is done on the whole object - I didn't test `.hashCode()`, but a pr is welcome! 😄
